kafka/client.go

```diff
@@ -0,0 +1,175 @@
+/*
+Copyright 2020 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package kafka
+
+import (
+	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"time"
+
+	"github.com/Shopify/sarama"
+	"github.com/kelseyhightower/envconfig"
+)
+
+type AdapterSASL struct {
+	Enable   bool   `envconfig:"KAFKA_NET_SASL_ENABLE" required:"false"`
+	User     string `envconfig:"KAFKA_NET_SASL_USER" required:"false"`
+	Password string `envconfig:"KAFKA_NET_SASL_PASSWORD" required:"false"`
+}
+
+type AdapterTLS struct {
+	Enable bool   `envconfig:"KAFKA_NET_TLS_ENABLE" required:"false"`
+	Cert   string `envconfig:"KAFKA_NET_TLS_CERT" required:"false"`
+	Key    string `envconfig:"KAFKA_NET_TLS_KEY" required:"false"`
+	CACert string `envconfig:"KAFKA_NET_TLS_CA_CERT" required:"false"`
+}
+
+type AdapterNet struct {
+	SASL AdapterSASL
+	TLS  AdapterTLS
+}
+
+type envConfig struct {
+	BootstrapServers []string `envconfig:"KAFKA_BOOTSTRAP_SERVERS" required:"true"`
+	Net              AdapterNet
+}
+
+func NewConsumer(ctx context.Context) (sarama.Client, error) {
+	cfg := sarama.NewConfig()
+	cfg.Version = sarama.V2_0_0_0
+
+	var env envConfig
+	if err := envconfig.Process("", &env); err != nil {
+		return nil, err
+	}
+
+	cfg.Consumer.Offsets.Initial = sarama.OffsetOldest
+	cfg.Consumer.Return.Errors = true
+
+	if env.Net.SASL.Enable {
+		cfg.Net.SASL.Enable = true
+		cfg.Net.SASL.User = env.Net.SASL.User
+		cfg.Net.SASL.Password = env.Net.SASL.Password
+	}
+
+	if env.Net.TLS.Enable {
+		cfg.Net.TLS.Enable = true
+		tlsConfig, err := newTLSConfig(env.Net.TLS.Cert, env.Net.TLS.Key, env.Net.TLS.CACert)
+		if err != nil {
+			return nil, err
+		}
+		cfg.Net.TLS.Config = tlsConfig
+	}
+
+	return sarama.NewClient(env.BootstrapServers, cfg)
+}
+
+func NewProducer(ctx context.Context) (sarama.Client, error) {
```

Comment:
To me these two functions looks more or less the same... what if this code more than having `NewConsumer`/`NewProducer`, has a `NewConfig` that then the KafkaSource or any other binding user can eventually customize?

Reply:
I factored a `NewConfig` with much thinner versions of these functions as "sugar".